src/sentry/scripts/similarity/index.lua

```diff
@@ -0,0 +1,618 @@
+groups that contain many different characteristic sets.
+
+This is modeled as two data structures:
+
+- A bucket frequency hash, which maintains a count of what buckets
+    have been recorded -- and how often -- in a ``(band, key)`` pair. This
+    data can be used to identify what buckets a key is a member of, and also
+    used to identify the degree of bucket similarity when comparing with data
+    associated with another key.
+- A bucket membership set, which maintains a record of what keys have been
+    record in a ``(band, bucket)`` pair. This data can be used to identify
+    what other keys may be similar to the lookup key (but not the degree of
+    similarity.)
+
+]]--
+local function range(start, stop)
+    local result = {}
+    for i = start, stop do
+        table.insert(result, i)
+    end
+    return result
+end
+
+function table.ifilter(t, f)
+    local result = {}
+    for i, value in ipairs(t) do
+        if f(value) then
+            table.insert(result, value)
+        end
+    end
+    return result
+end
+
+function table.imap(t, f)
+    local result = {}
+    for i, value in ipairs(t) do
+        result[i] = f(value)
+    end
+    return result
+end
+
+function table.ireduce(t, f, i)
+    if i == nil then
+        i = {}
+    end
+    local result = i
+    for i, value in ipairs(t) do
+        result = f(result, value)
+    end
+    return result
+end
+
+function table.izip(...)
+    local args = {...}
+    local n = math.max(
+        unpack(
+            table.imap(
+                args,
+                function (t) return #t end
+            )
+        )
+    )
+    local result = {}
+    for i = 1, n do
+        local item = {}
+        for j = 1, #args do
+            table.insert(item, args[j][i])
+        end
+        table.insert(result, item)
+    end
+    return result
+end
+
+function table.slice(t, start, stop)
+    -- NOTE: ``stop`` is inclusive!
+    local result = {}
+    for i = start or 1, stop or #t do
+        table.insert(result, t[i])
+    end
+    return result
+end
+
+
+-- Argument Parsing and Validation
+
+local function parse_number(value)
+    local result = tonumber(value)
+    assert(result ~= nil, 'got nil, expected number')
+    return result
+end
+
+local function parse_integer(value)
+    local result = parse_number(value)
+    assert(result % 1 == 0, 'got float, expected integer')
+    return result
+end
+
+local function build_argument_parser(fields)
+    return function (arguments)
+        local results = {}
+        for i = 1, #fields do
+            local name, parser = unpack(fields[i])
+            local value = arguments[i]
+            local ok, result = pcall(parser, value)
+            if not ok then
+                error(string.format('received invalid argument for %q in position %s with value %q; %s', name, i, value, result))
+            else
+                results[name] = result
+            end
+        end
+        return results, table.slice(arguments, #fields + 1)
```

Comment:
Thanks, I figured this would be a lot easier to debug than just having assertions scattered around since you don't really get a useful stack trace if anything breaks.